lib/WebDriver/AbstractWebDriver.php

```diff
@@ -122,15 +124,30 @@ protected function curl($requestMethod, $command, $parameters = null, $extraOpti
 
         list($rawResult, $info) = ServiceFactory::getInstance()->getService('service.curl')->execute($requestMethod, $url, $parameters, $extraOptions);
 
+        // According to the spec:
+        // - valid responses always include: sessionId, status and value
+        //   (at least as long as 'they include a message body' - but we tested commands such as timeouts/async_script
+        //   and session/delete and they still give back a json payload)
+        // - failed commands (http 500) include those as well
+        // - invalid requests (http 4xx) are instead plaintext
+        // See https://w3c.github.io/webdriver/webdriver-spec.html for details
+
+        if ($info['http_code'] >= 400 && $info['http_code'] <= 499) {
+            throw WebDriverException::factory(Exception::CURL_EXEC, 'Webdriver http error: ' . $info['http_code'] . ', payload :' . substr($rawResult, 0, 1000)/*, $previousException*/);
+        }
+
         $result = json_decode($rawResult, true);
-        $value   = null;
 
-        if (is_array($result) && array_key_exists('value', $result)) {
+        if (!is_array($result) || !array_key_exists('status', $result)) {
```

Comment:
Json without status is not sent back from Selenium, at least in my cursory testing.
But I wondered about the rare case where an intermediate proxy would also send back a valid json, or otherwise mess with the response, and thought that extra safety is good.

In fact in the curl info we have the content-type as well, so I might:
- avoid json-decode altogether if wrong content-type is found, and give a specific error msg
- use json_last_error in case json decoding fails to have the best possible error message